### ISSUE TITLE
Should there be a repository for "official" JUnit 5 extensions

### ISSUE THREAD

**Issue by smoyer64** [junit-team/junit5#321]
Many JUnit 5 Extensions will be lightweight glue code that connect two disparate projects.  Examples include the MockitoExtension and the JUnitParamsExtension (which is part of my AfterDiscoveryCallback PoC).  Other Extensions will replace existing JUnit 4 @Rules, such as the TemporaryFolder rule.  This class of extensions logically are only dependent upon JUnit 4.

Would it make sense to create a junit5-extensions project (repository) where members of the various project communities could collaborate?  When posed to the Mockito team in issue https://github.com/mockito/mockito/issues/438#issuecomment-227016382, @TimvdLippe suggested that the JUnit 5 team be asked.

**Comment by stefanbirkner** [comment]
I think that this doesn't make sense. The purpose of extension points is that independent projects can provide JUnit extensions independent from the JUnit team. JUnit should be the base framework that is extended by other libraries.

**Comment by raphw** [comment]
I think that the Mockito extension would follow the Mockito release-cycle so in a way it would of course make sense to add a `mockito-junit5` module. Currently, we already add several JUnit 4 runners backed by an optional dependency.

The problem I see currently is that we are still at Java 6 compatibility and will probably stay there for a while and few libraries already made this jump. From this perspective, we would need to add a new repository where releases are separate anyways. Therefore, I would still argue that a dedicated repository for JUnit extensions could make sense, also as a point of orientation for people looking for JUnit extensions.

**Comment by stefanbirkner** [comment]
Imagine that we have a dedicated repository for JUnit extensions (`junit-team/junit-extensions`). This repository will contain a lot of extensions. Therefore it has to be released whenever a single extension needs a new release. This workload can hardly be managed by the JUnit team.

Also from a user perspective it is a little bit confusing. You will have a lot of releases of e.g. `junit-mockito-extension` that are the same just because another extension has been changed. I think it is far easier to have a repository `mockito/junit-mockito-extension`. This has the advantage that the Mockito team has the control over releasing the extension.

**Comment by sbrannen** [comment]
> Would it make sense to create a junit5-extensions project (repository) where members of the various project communities could collaborate?

In short, no.

As a policy, JUnit projects do not depend on any third-party libraries or frameworks [*], and [for very good reasons](https://github.com/junit-team/junit5/wiki/Core-Principles#minimize-dependencies-especially-third-party).

As @stefanbirkner pointed out, "_JUnit should be the base framework that is extended by other libraries._" This holds for simpler extensions like the proposed `MockitoExtension` as well as for more complex extensions like the `SpringExtension`.

If the JUnit team maintains a collection of extensions, such extensions would have to rely only on libraries available in the JDK. Examples of such general purpose extensions might cover support for temporary folders, timing, conditions based on system properties or environment variables, etc. 

In summary, any third-party wishing to develop an extension to JUnit must implement and maintain that extension separately.

Regards,

Sam

[*] To my knowledge there are currently only two exceptions to this policy: 1) JUnit 4 has a dependency on Hamcrest (which was later admitted to have been a mistake); 2) `junit-platform-console` depends on JOpt Simple, but the latter will likely be shaded into the JAR in order to hide this dependency from consumers of the console runner.

**Comment by sbrannen** [comment]
**Team Decision:** close this issue as "won't fix".
